middleware/rate_limiter_test.go

```diff
@@ -0,0 +1,354 @@
+			return c.String(http.StatusOK, "test")
+		}
+
+		testCases := []struct {
+			id   string
+			code int
+		}{
+			{"127.0.0.1", http.StatusOK},
+			{"127.0.0.1", http.StatusOK},
+			{"127.0.0.1", http.StatusOK},
+			{"127.0.0.1", http.StatusTooManyRequests},
+			{"127.0.0.1", http.StatusTooManyRequests},
+			{"127.0.0.1", http.StatusTooManyRequests},
+			{"127.0.0.1", http.StatusTooManyRequests},
+		}
+
+		for _, tc := range testCases {
+			req := httptest.NewRequest(http.MethodGet, "/", nil)
+			req.Header.Add(echo.HeaderXRealIP, tc.id)
+
+			rec := httptest.NewRecorder()
+
+			c := e.NewContext(req, rec)
+			mw := RateLimiterWithConfig(RateLimiterConfig{
+				Store: &inMemoryStore,
+			})
+
+			_ = mw(handler)(c)
+
+			assert.Equal(t, tc.code, rec.Code)
+		}
+	}
+}
+
+func TestRateLimiterWithConfig_skipper(t *testing.T) {
+	e := echo.New()
+
+	var skipped bool
+	handler := func(c echo.Context) error {
+		skipped = true
+		return c.String(http.StatusOK, "test")
+	}
+	var inMemoryStore = NewRateLimiterMemoryStore(1, 3)
+
+	req := httptest.NewRequest(http.MethodGet, "/", nil)
+	req.Header.Add(echo.HeaderXRealIP, "127.0.0.1")
+
+	rec := httptest.NewRecorder()
+
+	c := e.NewContext(req, rec)
+
+	mw := RateLimiterWithConfig(RateLimiterConfig{
+		Skipper: func(c echo.Context) bool {
+			return true
+		},
+		Store: &inMemoryStore,
+		IdentifierExtractor: func(ctx echo.Context) (string, error) {
+			return "127.0.0.1", nil
+		},
+	})
+
+	_ = mw(handler)(c)
+
+	assert.Equal(t, true, skipped)
+}
+
+func TestRateLimiterWithConfig_beforeFunc(t *testing.T) {
+	e := echo.New()
+
+	handler := func(c echo.Context) error {
+		return c.String(http.StatusOK, "test")
+	}
+
+	var beforeRan bool
+	var inMemoryStore = NewRateLimiterMemoryStore(1, 3)
+
+	req := httptest.NewRequest(http.MethodGet, "/", nil)
+	req.Header.Add(echo.HeaderXRealIP, "127.0.0.1")
+
+	rec := httptest.NewRecorder()
+
+	c := e.NewContext(req, rec)
+
+	mw := RateLimiterWithConfig(RateLimiterConfig{
+		BeforeFunc: func(c echo.Context) {
+			beforeRan = true
+		},
+		Store: &inMemoryStore,
+		IdentifierExtractor: func(ctx echo.Context) (string, error) {
+			return "127.0.0.1", nil
+		},
+	})
+
+	_ = mw(handler)(c)
+
+	assert.Equal(t, true, beforeRan)
+}
+
+func TestRateLimiterMemoryStore_Allow(t *testing.T) {
+	var inMemoryStore = NewRateLimiterMemoryStore(1, 3, 2*time.Second)
+
+	testCases := []struct {
+		id      string
+		allowed bool
+	}{
+		{"127.0.0.1", true},
+		{"127.0.0.1", true},
+		{"127.0.0.1", true},
+		{"127.0.0.1", false},
+		{"127.0.0.1", false},
+		{"127.0.0.1", false},
+		{"127.0.0.1", false},
+	}
+
+	for _, tc := range testCases {
+		allowed := inMemoryStore.Allow(tc.id)
+		assert.Equal(t, tc.allowed, allowed)
+	}
+	/*
+	time.Sleep(5 * time.Second)
+	allowed := inMemoryStore.Allow("127.0.0.2")
+	assert.Equal(t, true, allowed)
+	//fmt.Println(inMemoryStore.visitors)
+	_, exists := inMemoryStore.visitors["127.0.0.1"]
+	assert.Equal(t, false, exists)
+	*/
```

Comment:
@pafuent @lammel 
I have been looking at this for an hour now and I honestly can't figure out why this is happening.
Could you help look at this commented block.

supposed to help test the cleanup using a time based approach.

Whenever I uncomment line 291, the test passes but when I leave the line commented the test fails.

You can recreate it if you'd like. 
Is there some hidden behavior in Golang causing this that I don't know about?

basically the `Allow` method calls `cleanupStaleVisitors` as a goroutine. Initially I thought it was the goroutine not being completed in time but when I increase the time to 7 secs it still gives the same behavior.

I need this to hit a 100% coverage (which you've probably noticed I love hitting 😅)

Reply:
I'll take a look.

Reply:
Please take a look at a branch in fork (didn't want to push here without asking):
* [example of producer to use for tests](https://github.com/lammel/echo/commit/27e7115c2228f656f6ae6a0eb0d6837769ca01f2)
* [fix for race](https://github.com/lammel/echo/commit/b5165d4b5b4faf5788d695bc730cf591f93c2c9c)

The race fix causes the mutex to guard all accesses to store fields (at the expense of lock contention).
I can push here it you want and think that this could be a way to go

Reply:
Forgot to note... I used for detecting the race your tests using the `-race` option for go test:

```
go test -race -timeout 30s -count=1 -v -run  ^TestRateLimiterMemoryStore_Allow$ github.com/labstack/echo/v4/middleware
```

Reply:
So the producer commit looks very great and quite intuitive. I love it ❤
The fix for the race though, I noticed you made the call to `cleanupStaleVisitors` synchronous. Any particular reason for that?

It isn't exactly friendly since the cleanup happens during a rate limit check it might keep a user waiting for a response for a while because the function clean up isn't complete

Reply:
I'd pull your producer commit now. thanks for the help !!

Reply:
I think making the cleanup method async doesn't play well... I just noticed that without it being a goroutine it runs perfectly.

But we need it to be async so that visitor will not suffer the burden of waiting for clean up to complete before getting authorized

Reply:
@lammel turns out even with the `rateProducer` the tests still have unpredictable results. you can see here

https://github.com/labstack/echo/pull/1724/checks?check_run_id=1579944790

Reply:
Well you made it async, without using the mutex in the cleanup routine. You are playing with fire and got burnt again.

I made it sync, because using the same mutex in the goroutine will effectivly make it sync too. But of course the async code can be improved by locking only for short periods and doing the cleanup in batches (usually shards are used for that). It can also work by using a previous and latest shard that is being flipped for each expiry interval (flipping will clear the new latest, a value is looked up in latest and if not found in previous). This way clearing will happen automatically, sync and lock free. The price is the flip (clear previous and switch pointer to make it latest) and the fallback lookup into previous shard, which is quite cheap. 

So either just make it sync for now, or work with shards to reduce lock contention or make it lock free.

Reply:
guys have you measured at which load locking gets problematic? 100, 1k, 10k, 100k request per second? It would be probably more maintainable to just have disclaimer that. 
> above load 10 000 requests per second you have better off using specific ratelimiting library instead of this middleware